### PR TITLE
Add OpenAPI generation

### PR DESCRIPTION
**Pull request by juanvallejo**
Fixes https://github.com/kubernetes-incubator/service-catalog/issues/339

This patch updates the openapi generator:

This patch fixes seg fault in openapi-gen/main.go and adds
a make rule to create the ./bin/openapi-gen binary as part
of the build process.

Generated go file gets placed under `pkg/openapi/openapi_generated.go`

@pwittrock @pmorie

### CODE REVIEW

File: cmd/libs/go2idl/openapi-gen/main.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	"k8s.io/kubernetes/cmd/libs/go2idl/client-gen/types"
+	"k8s.io/kubernetes/cmd/libs/go2idl/openapi-gen/generators"
+
+	"github.com/golang/glog"
+)
+
+var (
+	inputVersions = []string{
+		"api/",
+		"authentication/",
```
Comment (juanvallejo) [inline cmd/libs/go2idl/openapi-gen/main.go]: @pmorie or @pwittrock not sure what paths this should be looking at. These were taken from the k8s client-gen

**Comment by MHBauer** [comment]
If this is another call on a binary from upstream, I'd like to compile and call it directly.

File: Makefile [hunk 2]
```diff
@@ -98,7 +98,8 @@ $(BINDIR)/apiserver: .init .generate_files cmd/apiserver $(NEWEST_GO_FILE)
 #################################################
 .generate_exes: $(BINDIR)/defaulter-gen \
                 $(BINDIR)/deepcopy-gen \
-                $(BINDIR)/conversion-gen
+                $(BINDIR)/conversion-gen \
+				$(BINDIR)/openapi-gen
```
Comment (duglin) [inline Makefile]: indent is off

File: cmd/libs/go2idl/openapi-gen/BUILD [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package(default_visibility = ["//visibility:public"])
```
Comment (pmorie) [inline cmd/libs/go2idl/openapi-gen/BUILD]: We don't need this file - our repo does not use bazel

File: cmd/libs/go2idl/openapi-gen/.import-restrictions [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+  "Rules": [
+    {
+      "SelectorRegexp": "k8s[.]io",
```
Comment (pmorie) [inline cmd/libs/go2idl/openapi-gen/.import-restrictions]: Dhoukd these be different for our repo?

File: cmd/libs/go2idl/openapi-gen/generators/openapi_test.go [hunk 5]
```diff
@@ -0,0 +1,337 @@
+/*
```
Comment (pmorie) [inline cmd/libs/go2idl/openapi-gen/generators/openapi_test.go]: Do we need to import this binary? Can we go build the upstream one?

File: cmd/libs/go2idl/openapi-gen/.import-restrictions (hunk 4 above)
Comment (juanvallejo) [inline cmd/libs/go2idl/openapi-gen/.import-restrictions]: Not sure what these should be

**Comment by ncdc** [comment]
I wouldn't recommend placing the openapi files in `pkg/client/listers`.

I'd recommend looking at https://github.com/openshift/origin/blob/master/hack/update-generated-openapi.sh for how OpenShift generates its openapi spec to include both Kubernetes and OpenShift api types.

This is OpenShift's main.go for it: https://github.com/openshift/origin/blob/master/tools/genopenapi/openapi.go

File: cmd/libs/go2idl/openapi-gen/main.go (hunk 1 above)
Comment (ncdc) [inline cmd/libs/go2idl/openapi-gen/main.go]: See my comment on the overall PR. If you follow the OpenShift setup, you won't need to do this in go. The shell script does a grep to set up the input dirs.

**Comment by juanvallejo** [comment]
@ncdc thanks for the link to the openshift script. I have updated this PR to re-use the vendored openapi-gen with it, however when testing it, I realized that our vendored `client-go` is not structured as `vendor/k8s.io/client-go/pkg/api...` but rather as `vendor/k8s.io/client-go/1.5/pkg/api...` thus causing runtime errors for the openapi-gen script when trying to look for client-go deps. Do we do anything different in openshift when vendoring client-go?

**Comment by ncdc** [comment]
@juanvallejo as client-go has advanced, its directory structure has changed. I don't know specifically what issue you're hitting, but you'll need to adjust things appropriately if you're getting import issues.

File: Makefile [hunk 6]
```diff
@@ -110,6 +111,9 @@ $(BINDIR)/deepcopy-gen: .init cmd/libs/go2idl/deepcopy-gen
 $(BINDIR)/conversion-gen: cmd/libs/go2idl/conversion-gen
 	$(DOCKER_CMD) go build -o $@ $(SC_PKG)/$^
 
+$(BINDIR)/openapi-gen: vendor/k8s.io/kubernetes/cmd/libs/go2idl/openapi-gen
+	$(DOCKER_CMD) go build -o $@ $(SC_PKG)/$^
```
Comment (MHBauer) [inline Makefile]: yay! 👍

File: contrib/hack/update-generated-openapi.sh [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    grep --color=never -rl '+k8s:openapi-gen=' pkg | \
+    xargs -n1 dirname | \
+    sed "s,^,${SC_PREFIX}," | \
+    sort -u
+  )
+)
+
+INPUT_DIRS=$(IFS=,; echo "${INPUT_DIRS[*]}")
+
+genopenapi \
```
Comment (MHBauer) [inline contrib/hack/update-generated-openapi.sh]: what is this binary?

**Comment by MHBauer** [comment]
I was on a roll, so I whipped this out, https://github.com/MHBauer/service-catalog/commit/392a1ac812a20709edf8786271c8b555c22eefe1

It doesn't succeed. I get a panic:
```

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x2b8 pc=0x4c6165]

goroutine 1 [running]:
panic(0x873e40, 0xc420014070)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/gengo/parser.(*Builder).importBuildPackage(0xc42032c9b0, 0x7fff8a67ae00, 0x42, 0x20, 0x6, 0x1)
	/go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/gengo/parser/parse.go:137 +0x205
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/gengo/parser.(*Builder).addDir(0xc42032c9b0, 0x7fff8a67ae00, 0x42, 0x1, 0xb6ff60, 0xc425f58700)
	/go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/gengo/parser/parse.go:245 +0x70
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/gengo/parser.(*Builder).AddDirTo(0xc42032c9b0, 0x7fff8a67ae00, 0x42, 0xc4256ca7e8, 0x50, 0xc425f58780)
	/go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/gengo/parser/parse.go:231 +0x11f
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/gengo/generator.(*Context).AddDir(0xc4256ca7e0, 0x7fff8a67ae00, 0x42, 0x280, 0x2a6)
	/go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/gengo/generator/generator.go:210 +0x4c
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/cmd/libs/go2idl/openapi-gen/generators.Packages(0xc4256ca7e0, 0xc42049d300, 0x8f04b5, 0x3, 0xc4256ca7e0)
	/go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/cmd/libs/go2idl/openapi-gen/generators/openapi.go:95 +0x26b
github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/gengo/args.(*GeneratorArgs).Execute(0xc42049d300, 0xc4202f9620, 0x8f04b5, 0x3, 0x9308f8, 0x40e479, 0xc420010118)
	/go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/gengo/args/args.go:165 +0x170
main.main()
	/go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes/cmd/libs/go2idl/openapi-gen/main.go:43 +0x234
```

**Comment by MHBauer** [comment]
I figured it out, the package you are generating into has to 'exist' by having a go file in it already.

compare https://github.com/kubernetes-incubator/service-catalog/compare/master...MHBauer:vendored-openapi-gen?expand=1

I don't know if that looks correct, but I do see our service catalog types.

@pwittrock @juanvallejo @pmorie @duglin

File: contrib/hack/update-generated-openapi.sh (hunk 7 above)
Comment (juanvallejo) [inline contrib/hack/update-generated-openapi.sh]: Oversight on my part when re-using this script from openshift, will fix: `s,genopenapi,./bin/openapi-gen,`. Will also ensure the binary exists / is built before getting to this point in the script

**Comment by juanvallejo** [comment]
@MHBauer 
> I figured it out, the package you are generating into has to 'exist' by having a go file in it already.

Yeah, I was running into this same error when trying this out earlier.
Maybe either have placeholder go files in the root of all packages that the script selects, or filter the list of packages to only those that have go files at the top level?

List of selected packages:
```
$ grep --color=never -rl '+k8s:openapi-gen=' vendor/k8s.io/kubernetes | \
    xargs -n1 dirname | \
    sed "s,^vendor/,," | \
    sort -u | xsel --input --clipboard

k8s.io/kubernetes
k8s.io/kubernetes/cmd/libs/go2idl/client-gen/test_apis/testgroup/v1
k8s.io/kubernetes/cmd/libs/go2idl/openapi-gen
k8s.io/kubernetes/federation/apis/federation/v1beta1
k8s.io/kubernetes/pkg/api/resource
k8s.io/kubernetes/pkg/apis/abac
k8s.io/kubernetes/pkg/apis/abac/v0
k8s.io/kubernetes/pkg/apis/abac/v1beta1
k8s.io/kubernetes/pkg/apis/apps
k8s.io/kubernetes/pkg/apis/apps/v1beta1
k8s.io/kubernetes/pkg/apis/authentication
k8s.io/kubernetes/pkg/apis/authentication/v1beta1
k8s.io/kubernetes/pkg/apis/authorization
k8s.io/kubernetes/pkg/apis/authorization/v1beta1
k8s.io/kubernetes/pkg/apis/autoscaling
k8s.io/kubernetes/pkg/apis/autoscaling/v1
k8s.io/kubernetes/pkg/apis/batch
k8s.io/kubernetes/pkg/apis/batch/v1
k8s.io/kubernetes/pkg/apis/batch/v2alpha1
k8s.io/kubernetes/pkg/apis/certificates
k8s.io/kubernetes/pkg/apis/certificates/v1alpha1
k8s.io/kubernetes/pkg/apis/componentconfig
k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1
k8s.io/kubernetes/pkg/apis/extensions
k8s.io/kubernetes/pkg/apis/extensions/v1beta1
k8s.io/kubernetes/pkg/apis/imagepolicy
k8s.io/kubernetes/pkg/apis/imagepolicy/v1alpha1
k8s.io/kubernetes/pkg/apis/meta/v1
k8s.io/kubernetes/pkg/apis/policy
k8s.io/kubernetes/pkg/apis/policy/v1beta1
k8s.io/kubernetes/pkg/apis/rbac
k8s.io/kubernetes/pkg/apis/rbac/v1alpha1
k8s.io/kubernetes/pkg/apis/storage/v1beta1
k8s.io/kubernetes/pkg/api/v1
k8s.io/kubernetes/pkg/genericapiserver/openapi
k8s.io/kubernetes/pkg/runtime
k8s.io/kubernetes/pkg/util/intstr
k8s.io/kubernetes/pkg/version
k8s.io/kubernetes/pkg/watch/versioned
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/api/resource
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/apps
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/apps/v1beta1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/authentication
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/authentication/v1beta1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/authorization
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/authorization/v1beta1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/autoscaling
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/autoscaling/v1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/batch
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/batch/v1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/batch/v2alpha1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/certificates
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/certificates/v1alpha1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/componentconfig
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/componentconfig/v1alpha1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/extensions
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/extensions/v1beta1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/imagepolicy
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/imagepolicy/v1alpha1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/meta/v1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/policy
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/policy/v1beta1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/rbac
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/rbac/v1alpha1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/apis/storage/v1beta1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/api/v1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/federation/apis/federation/v1beta1
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/runtime
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/util/intstr
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/version
k8s.io/kubernetes/staging/src/k8s.io/client-go/pkg/watch/versioned
```

**Comment by juanvallejo** [comment]
Hm, after rebasing and re-running `./contrib/hack/update-generated-openapi.sh`, I'm now getting the error:
```
F0209 01:14:17.389032    6159 main.go:44] Error: Failed making a parser: unable to add directory "k8s.io/kubernetes": unable to *find* "k8s.io/kubernetes": cannot find package "k8s.io/kubernetes" in any of:
        /home/jvallejo/go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/kubernetes (vendor tree)
        /usr/local/go/src/k8s.io/kubernetes (from $GOROOT)
        /home/jvallejo/go/src/k8s.io/kubernetes (from $GOPATH)
```

**Comment by ncdc** [comment]
@juanvallejo if the servicecatalog api types depend on kube api types (I assume they will), you will need to include the kube api types in --input-dirs

**Comment by MHBauer** [comment]
Probably need to run that script in our build container. @duglin @krancour thoughts?

**Comment by MHBauer** [comment]
@ncdc Why do we need any of the API information other than our own? We're not serving the core k8s API.

**Comment by juanvallejo** [comment]
@ncdc thanks for the feedback, that was the part I was having some trouble with: some of the paths generated here: https://github.com/kubernetes-incubator/service-catalog/pull/349#issuecomment-278556609 did not have go files at the top-level, which caused https://github.com/kubernetes-incubator/service-catalog/pull/349#issuecomment-278534791 to happen

**Comment by ncdc** [comment]
@MHBauer because the spec you generate is going to need to reference `ObjectMeta`, `ListMeta`, and other things imported from kube.

**Comment by MHBauer** [comment]
That should be exactly only `"k8s.io/kubernetes/pkg/api/v1"` & `"k8s.io/kubernetes/pkg/apis/meta/v1"` in our case?

**Comment by ncdc** [comment]
Yeah, whatever you're actually importing and using

**Comment by MHBauer** [comment]
@ncdc updated, https://github.com/kubernetes-incubator/service-catalog/compare/master...MHBauer:vendored-openapi-gen

I think it pulled in a lot of other junk, but it has typemeta and objectmeta now. 

Do we need the openshift script still?

**Comment by ncdc** [comment]
Not if it's working for you as-is, no

**Comment by juanvallejo** [comment]
@ncdc @MHBauer Thanks for the feedback and help. I have updated the Makefile to include both k8s paths in https://github.com/kubernetes-incubator/service-catalog/pull/349#issuecomment-278681835

pkg/openapi/openapi_generated.go now matches the one in https://github.com/kubernetes-incubator/service-catalog/compare/master...MHBauer:vendored-openapi-gen

Are we tracking the generated file as well?

**Comment by MHBauer** [comment]
unfortunately, yes. Usually done as two commits, one for the code that does the generation, one for the generated code.

**Comment by MHBauer** [comment]
Is there anything we can do to *know* it's working? 

As far as you are aware the output looks correct?

File: Makefile [hunk 8]
```diff
@@ -166,6 +170,12 @@ $(BINDIR)/informer-gen:
 		--input-dirs "github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1alpha1" \
 		--listers-package "github.com/kubernetes-incubator/service-catalog/pkg/client/listers" \
 		--output-package "github.com/kubernetes-incubator/service-catalog/pkg/client/informers"
+	# generate openapi
+	$(DOCKER_CMD) $(BINDIR)/openapi-gen \
+		--v 1 --logtostderr \
+		--go-header-file "vendor/github.com/kubernetes/repo-infra/verify/boilerplate/boilerplate.go.txt" \
+		--input-dirs "github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog/v1alpha1,k8s.io/kubernetes/pkg/api/v1,k8s.io/kubernetes/pkg/apis/meta/v1" \
+		--output-package "github.com/kubernetes-incubator/service-catalog/pkg/openapi"
```
Comment (MHBauer) [inline Makefile]: Is this the standard "openapi" package that upstream k8s uses? We want to match.

Comment (ncdc) [inline Makefile]: It doesn't really matter

**Comment by juanvallejo** [comment]
@MHBauer 
> Is there anything we can do to know it's working?

maybe a test discovery client that uses `https://localhost:6443/apis/servicecatalog.k8s.io/v1alpha1/...` and is used to determine which types are supposed to have generated documentation by openapi-gen?

**Comment by pmorie** [comment]
LGTM assuming the CI failures are fixed; thanks for doing this

File: Makefile [hunk 9]
```diff
@@ -122,6 +123,9 @@ $(BINDIR)/lister-gen:
 $(BINDIR)/informer-gen:
 	$(DOCKER_CMD) go build -o $@ $(SC_PKG)/vendor/k8s.io/kubernetes/cmd/libs/go2idl/informer-gen
 
+$(BINDIR)/openapi-gen: vendor/k8s.io/kubernetes/cmd/libs/go2idl/openapi-gen
```
Comment (MHBauer) [inline Makefile]: oh cool. I didn't realize directories can be targets.

**Comment by arschles** [comment]
LGTM, merging

**Pull request merged**